core/src/main/java/com/graphhopper/routing/Router.java

```diff
@@ -373,6 +374,11 @@ private void checkMaxVisitedNodes() {
                 throw new IllegalArgumentException("The max_visited_nodes parameter has to be below or equal to:" + routerConfig.getMaxVisitedNodes());
         }
 
+        private void checkTimeout() {
+            if (getTimeoutMillis(request.getHints()) > routerConfig.getTimeoutMillis())
```

Comment:
I would prefer to take the `Math.min()` approach instead. Lowering the timeout on the serverside would cause all requests to fail which probably isn't expected.

Reply:
Not sure tbh, we could even make this configurable :)

Reply:
Ok, you mean there should not be an immediate error when we request a longer timeout than what is specified on the server-side, but the server-side timeout should be used silently?

Reply:
Ok, guessing and trial and error would only be needed if the server limit was not documented. But I see your point that such errors aren't very helpful. 

Actually I am wondering why a 'client' would setup a timeout in the first place, especially if it is not allowed to request a **higher** limit. When the GH server is used without a proxy only the server limit seems useful, because clients do not benefit from setting a limit at all. When the GH server is used with a proxy the best place to set the limit would be the proxy in which case the server limit is not needed. I don't really see a scenario where a server limit will be used and clients set the timeout as well?

Reply:
> When the GH server is used with a proxy the best place to set the limit would be the proxy in which case the server limit is not needed.

Is it? The gist of #2349 was that we're unable to tell if the client or proxy cancelled the request.

Reply:
For me, the client limit is more of a hint, and it is up to the server to enforce a reasonable limit.

Reply:
> Is it? The gist of https://github.com/graphhopper/graphhopper/issues/2349 was that we're unable to tell if the client or proxy cancelled the request.

Yes, I mean the value of the timeout parameter would be specified by the proxy, but of course the timeout is implemented in the GH server (this PR).

Reply:
> For me, the client limit is more of a hint, and it is up to the server to enforce a reasonable limit.

Under which circumstances would a client like to hint for a **shorter** timeout?

Reply:
> Under which circumstances would a client like to hint for a **shorter** timeout?

I can see the more theoretical use case of a request that needs to be completed within a certain timeframe from the perspective of the client, but I agree that this limit is more suited to server administrators who want to restrict resource usage.

"I don't care about the result if the calculation takes longer than x milliseconds" vs "I don't want client requests to take longer than x milliseconds".

> Yes, I mean the value of the timeout parameter would be specified by the proxy, but of course the timeout is implemented by the GH server (this PR).

It just seems a bit odd to introduce a parameter that's only intended for proxies or other middleware. The worst case scenario would be for a user to mistakenly use this parameter instead of a simple HTTP timeout. Things would work at first glance, until the admin decides to introduce a server-side limit via configuration.

My only intention here is to avoid a potential footgun. Using the lower of both limits would work for both cases.

Reply:
> I can see the more theoretical use case of a request that needs to be completed within a certain timeframe from the perspective of the client

But then the client simply uses a timeout on the client side, no?

> but I agree that this limit is more suited to server administrators who want to restrict resource usage.

I think that's the only use-case. Nobody forces any client to wait until the server responds or times out. Any client can move on whenever it wants to.

> My only intention here is to avoid a potential footgun. Using the lower limit would work for both cases.

Same here :) To me doing something silently that is contrary to what was requested is normally a no-go. Like using a timeout of 1000ms, when a request came in requesting `timeout_ms=5000`. I'd much rather return an error in this case and handle the error where it occurred (at the place the timeout is set). I agree that if that place is the **end**-user such an error can be annoying, but like I said, I don't see a reason why an end-user should setup a timeout in the first place.

Ok, but using the minimum we can support both scenarios: Standalone servers can setup the limit on the server-side only and clients will probably never use the timeout parameter, but even if they do they cannot really do any harm.
In case a proxy is used the server limit can be set to the maximum (or left out since this is the default) and the timeout can be controlled via the request parameter.

This comes at the cost of the (to me unexpected) behavior, that the server can ignore the requested timeout and use a lower limit silently.

Reply:
>  Standalone servers can setup the limit on the server-side only and clients will probably never use the timeout parameter, but even if they do they cannot really do any harm.

Hm, but if they never use it there is no real problem to thrown an exception when it is above the limit? ;)

> Things would work at first glance, until the admin decides to introduce a server-side limit via configuration.

I think my point is that end-users will never set a timeout (why would they?), so changing the value on the server-side won't cause any problems.

Reply:
I think it's fine either way. But the parameter should be properly documented to reflect that.

Reply:
Ok, I do not care very much, either. I'll use the minimum for now.